File: contest/LeetCode_Weekly_Contest_201/Maximum_Number_of_Non-Overlapping_Subarrays_With_Sum_Equals_Target.cpp
const int N = 100005;
int jp[N];
int dp[N];

class Solution {
public:
    int maxNonOverlapping(vector<int>& nums, int target) {
        int n = nums.size();
        memset(jp, -1, sizeof jp);
        memset(dp, -1, sizeof dp);
        map<int, int> pos;
        
        dp[n] = 0;
        pos[0] = n;
        
        int ret = 0;
        int suffix = 0;
        for (int i = n-1; i >= 0; --i) {
            dp[i] = dp[i+1];
            suffix += nums[i];
            int v = suffix - target;
            if (pos.count(v)) {
                int p = pos[v];
                dp[i] = max(dp[i], dp[p] + 1);    
            }
            pos[suffix] = i;
            ret = max(ret, dp[i]);
        }
        return ret;
    }
};

File: contest/LeetCode_Weekly_Contest_201/Maximum_Number_of_Non-Overlapping_Subarrays_With_Sum_Equals_Target.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var maxNonOverlapping = function(nums, target) {
    var n = nums.length;
    var dp = [];
    for (var i = 0; i <= n; ++i) dp[i] = 0;
    var pos = {0: n};
    var suffix = 0;
    var ret = 0;
    
    for (var i = n-1; i >= 0; --i) {
        dp[i] = Math.max(dp[i], dp[i+1]);
        suffix += nums[i];
        var v = suffix - target;
        if (pos.hasOwnProperty(v)) {
            var p = pos[v];
            dp[i] = Math.max(dp[i], dp[p] + 1);
        }
        pos[suffix] = i;
        ret = Math.max(ret, dp[i]);
    }
    return ret;
};
